nova/objects/resource_provider.py

```diff
@@ -1832,6 +1832,21 @@ def _check_capacity_exceeded(ctx, allocs):
         raise exception.InvalidInventory(resource_class=class_str,
                 resource_provider=provider_str)
 
+    # check if this a complete replacement of one consumer with another one
+    # TODO(jkulik): Make this only work if it's replacing to same or less.
+    consumer_resources = collections.defaultdict(list)
+    for alloc in allocs:
+        d = (alloc.resource_class, alloc.used)
+        consumer_resources[alloc.consumer_id].append(d)
+    is_replacement = len(consumer_resources) == 2
+    if is_replacement:
+        for items in six.moves.zip_longest(*consumer_resources.values(),
+                                           fillvalue=(None, None)):
```

Comment:
Isn't what you want `itertools.product`? Or put another way: Who's to say that `customer_resources.values()` isn't "(-cpu, +ram), (+cpu, -ram)"?

Reply:
Btw, this whole hunk took me 15minutes to properly decipher in my head... I tried but also couldn't readily find any more readable way to do it, but it's still not readable to me.

Reply:
Thank you for your feedback. This is actually a good case I didn't test/think about. I assume resources are always put in in the same order, but nevertheless added a `sort()` step and adapted the tests to put in the second resources in another order.

Additionally, I added some more comments. Maybe this helps in the future.